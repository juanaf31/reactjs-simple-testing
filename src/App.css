.wrapper{
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
}

.hovered{
  background-color: blue;
  color: white;
}