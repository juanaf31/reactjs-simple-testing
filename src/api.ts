export const api = {
    createItem: (url:any,newItem:any) => {
      return Promise.resolve(newItem);
    }
  };