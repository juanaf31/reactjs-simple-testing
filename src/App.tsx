import React, { useState } from 'react';
import './App.css';
import {api} from './api'

interface IItems{
  id:number,
  item:string
}

function App() {
  const [items,setItems]=useState<IItems[]>([])
  const [text,setText]=useState('')
  const [isHovered, setIsHovered]=useState(false)
  const handleSubmit=(e:any)=>{
    e.preventDefault()
    const newItem = {id:items.length+1,item:text}
    api.createItem('/item',newItem).then((res)=>{
      console.log(res)
      setItems((prev)=>[...prev,res])
      setText('')
    }
    )
  }
  const handleChange=(e:any)=>{
    setText(e.target.value)
  }
  const handleClear=()=>{
    setItems([])
  }
  return (
    <div className='App'>
      <div className='wrapper'>
        <h1>Nusantech TODOs</h1>
      
        <form onSubmit={handleSubmit}>
            <label htmlFor="data-input">What you want to do?</label>
            <br/>
            <input id='data-input' onChange={handleChange} value={text}/>
            <button data-testid='buttonSubmit' className={isHovered?'hovered':''} onMouseEnter={()=>setIsHovered(true)} onMouseLeave={()=>setIsHovered(false)}>
              {`Add #${items.length+1}`}
            </button>
            <button type='button' onClick={handleClear}>clear</button>
        </form>
        <ul>
          {items.map((item)=>(
            <li key={item.id}>{item.item}</li>
            ))}
        </ul>
      </div>
    </div>
  );
}

export default App;
