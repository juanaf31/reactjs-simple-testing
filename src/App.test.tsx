import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import App from './App';
import renderer from 'react-test-renderer'

import {api} from './api'

const mockCreateItem = (api.createItem=jest.fn())

test('renders app component', () => {
  const { getByText, getByLabelText } = render(<App/>);

  getByText('Nusantech TODOs')
  getByLabelText('What you want to do?')
  getByText('Add #1')
  getByText('clear')
});

test('Add style when hover',()=>{
  const {getByTestId}=render(<App/>)

  const button = getByTestId('buttonSubmit')

  fireEvent.mouseEnter(button)
  expect(button).toHaveClass('hovered')
})

test('Remove style when mouse leave',()=>{
  const {getByTestId}=render(<App/>)

  const button = getByTestId('buttonSubmit')
  
  fireEvent.mouseLeave(button)
  expect(button).not.toHaveClass()
})

test('makes sure input value correct',()=>{
  const {getByLabelText}=render(<App/>)

  const input = getByLabelText('What you want to do?') as HTMLInputElement

  fireEvent.change(input, {target:{value:'go to school'}})

  expect(input.value).toBe('go to school')
  expect(input.value).not.toBe('go to schools')
})

test('allows user to add list',async ()=>{
  const todo = 'This is Todo'
  mockCreateItem.mockResolvedValueOnce({id:1,item:todo})
  
  const {getByText,getByLabelText, getByTestId}=render(<App/>)

  const input = getByLabelText('What you want to do?')
  const button = getByTestId('buttonSubmit')

  fireEvent.change(input,{target:{value:todo}})

  fireEvent.click(button)
  
  expect(mockCreateItem).toHaveBeenCalledTimes(1)
  expect(mockCreateItem).toHaveBeenCalledWith(
    "/item",
    {id:1,item:todo}
    )
    await waitFor(()=>getByText(todo))
  expect(button.textContent).toBe('Add #2')
})

test('allows user to clear list',()=>{
  const {getByText,getByLabelText, getByTestId}=render(<App/>)

  const input = getByLabelText('What you want to do?')
  const button = getByText('clear')
  const buttonSubmit = getByTestId('buttonSubmit')

  fireEvent.click(button)
  expect(input.textContent).toBe('')
  expect(buttonSubmit.textContent).toBe('Add #1')
})


test('button matches snapshot',()=>{
  const tree = renderer.create(<App/>).toJSON()
  expect(tree).toMatchSnapshot()
})