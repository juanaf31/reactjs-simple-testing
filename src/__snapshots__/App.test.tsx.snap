// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button matches snapshot 1`] = `
<div
  className="wrapper"
>
  <h1>
    Nusantech TODOs
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="data-input"
    >
      What you want to do?
    </label>
    <br />
    <input
      id="data-input"
      onChange={[Function]}
      value=""
    />
    <button
      className=""
      data-testid="buttonSubmit"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      Add #1
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      clear
    </button>
  </form>
  <ul />
</div>
`;
